Enum of ResponseContext keys

esevastyanov opened the pull request:
### Description
Aggregated `ResponseContext` keys into enum as a next step of [ResponseContext refactoring](https://github.com/apache/incubator-druid/pull/8110). Previously the keys were just static strings so theoretically there was no obstacle to use any string as a key of `ReponseContext`. This refactoring eliminates this possibility by introducing the enum of `ResponseContext` keys and exposing only methods requiring the enum instance as a key.
#### Fixed the issue of merging `ResponseContext` instances returned by Historicals to Broker
There was no rule of merging different response contexts but from my point of view, the current solution of rewriting existing values by last ones is incorrect because of losing valuable information. For example, a value associated with the key `UNCOVERED_INTERVALS` contains a list of uncovered intervals and the result value is not the last returned but the concatenation of all returned lists. The same issue with the key `MISSING_SEGMENTS` (a list of missing segments) and `COUNT` (the number of scanned rows). Thus I decided to provide every key with a merge function. So response contexts merge became a simple procedure.

Also improved `ResponseContext` serialization. Previously the result of serialization was truncated if its length was greater than the limit. I believe it would be better to keep the context structure and make it deserializable thus the process of serialization removes max-length fields completely from the context until the final result length doesn't exceed the limit.
<hr>

This PR has:
- [x] been self-reviewed.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [x] added unit tests or modified existing tests to cover new code paths.

<hr>

For reviewers: the key changed classes is `ResponseContext`.

leventov commented on processing/src/main/java/org/apache/druid/query/context/ResponseContext.java:
```diff
@@ -76,56 +138,128 @@ public static ResponseContext createEmpty()
   }
 
-  public Object get(String key)
+  protected abstract Map<String, Object> getDelegate();
```
Please add a Javadoc comment explaining why the keys should still be Strings rather than Keys. (Or change, if they don't.)

esevastyanov replied:
This requires updating SerDe (type reference and commenting that object mapper should have enum support, considering we have custom value name) or converting map keys to strings on serialization and back on deserialization.

leventov replied:
Could you please add a comment like `we are mapping from Strings rather than {@link BaseKey}s because ...`?

esevastyanov replied:
Updated the delegate type to `Map<BaseKey, Object>`

leventov commented on processing/src/main/java/org/apache/druid/query/scan/ScanQueryEngine.java:
```diff
@@ -266,7 +261,7 @@ public void cleanup(Iterator<ScanResultValue> iterFromMake)
   private long calculateLimit(ScanQuery query, ResponseContext responseContext)
```
I think it would be clearer to call this method "calculateRemainingScanRowsLimit"

esevastyanov replied:
Renamed

leventov commented on processing/src/main/java/org/apache/druid/query/scan/ScanQueryEngine.java:
```diff
@@ -266,7 +261,7 @@ public void cleanup(Iterator<ScanResultValue> iterFromMake)
-      return query.getLimit() - (long) responseContext.get(ResponseContext.CTX_COUNT);
+      return query.getLimit() - (long) responseContext.get(ResponseContext.Key.COUNT);
     }
     return query.getLimit();
```
Please rename this property to "scanRowsLimit" for clarity.

esevastyanov replied:
Renamed

leventov commented on server/src/main/java/org/apache/druid/client/CachingClusteredClient.java:
```diff
@@ -358,8 +358,8 @@ private void computeUncoveredIntervals(TimelineLookup<String, ServerSelector> ti
         // Which is not necessarily an indication that the data doesn't exist or is
```
The phrase above "This returns intervals..." is strange. I would say "Record in the response context the intervals..."

esevastyanov replied:
Updated

leventov commented on server/src/main/java/org/apache/druid/server/QueryResource.java:
```diff
@@ -252,24 +252,30 @@ public void write(OutputStream outputStream) throws WebApplicationException
-        if (responseContext.get(ResponseContext.CTX_ETAG) != null) {
-          builder.header(HEADER_ETAG, responseContext.get(ResponseContext.CTX_ETAG));
-          responseContext.remove(ResponseContext.CTX_ETAG);
+        if (responseContext.get(ResponseContext.Key.ETAG) != null) {
```
Double get looks awkward. It could be
```java
Object entityTag = responseContext.remove(ResponseContext.Key.ETAG);
if (entityTag != null) {
  builder.header(HEADER_ETAG, entityTag);
}
```

esevastyanov replied:
Nice catch, updated

leventov commented on server/src/main/java/org/apache/druid/server/QueryResource.java:
```diff
@@ -252,24 +252,30 @@ public void write(OutputStream outputStream) throws WebApplicationException
-          builder.header(HEADER_ETAG, responseContext.get(ResponseContext.CTX_ETAG));
-          responseContext.remove(ResponseContext.CTX_ETAG);
+        if (responseContext.get(ResponseContext.Key.ETAG) != null) {
+          builder.header(HEADER_ETAG, responseContext.get(ResponseContext.Key.ETAG));
```
I think would be clearer to call this variable `responseBuilder`

esevastyanov replied:
Renamed

leventov commented on server/src/main/java/org/apache/druid/server/QueryResource.java:
```diff
@@ -252,24 +252,30 @@ public void write(OutputStream outputStream) throws WebApplicationException
+            RESPONSE_CTX_HEADER_LEN_LIMIT
+        );
+        if (serializationResult.isReduced()) {
+          log.warn(
```
Should Druid cluster operators monitor these messages? Can they do anything about them? If not, this should probably be `info()`. See #7362.

esevastyanov replied:
I'm not sure about that, even left the log message as is although the context is not truncated anymore but "reduced". According to the [corresponding PR discussion](https://github.com/apache/incubator-druid/pull/2336), it was important to have a log message with full context, it's likely someone has a filter in a log aggregator for this kind of message.
BTW I see your point and started discussion to only mention backward compatibility for log filters.

leventov replied:
Please add a comment like "Whether or not this logging statement should properly be on the WARN level (which is unclear), it's kept on the warn level for backward compatibility: see https://github.com/apache/incubator-druid/pull/2336".

(If I understood your comment correctly.)

esevastyanov replied:
Since the change will be tagged as `incompatible` I decided to update the log level to `info`

leventov commented:
Labelled `ResponseContext` because this PR heavily affects `ResponseContext`, a `@PublicApi` class.

esevastyanov commented on processing/src/main/java/org/apache/druid/query/context/ResponseContext.java:
```diff
@@ -21,51 +21,113 @@
+     */
+    UNCOVERED_INTERVALS(
+        "uncoveredIntervals",
+            (oldValue, newValue) -> {
```
Right, we need to fix checkstyle config

esevastyanov replied:
Looks like this cannot be fixed just by updating the config as checkstyle plugin has some known issues with indentation of lambdas as arguments
https://github.com/checkstyle/checkstyle/issues/4638
https://github.com/checkstyle/checkstyle/issues/3342

esevastyanov commented on server/src/main/java/org/apache/druid/client/CachingClusteredClient.java:
```diff
@@ -358,8 +358,8 @@ private void computeUncoveredIntervals(TimelineLookup<String, ServerSelector> ti
-        responseContext.put(ResponseContext.CTX_UNCOVERED_INTERVALS, uncoveredIntervals);
-        responseContext.put(ResponseContext.CTX_UNCOVERED_INTERVALS_OVERFLOWED, uncoveredIntervalsOverflowed);
+        responseContext.merge(ResponseContext.Key.UNCOVERED_INTERVALS, uncoveredIntervals);
+        responseContext.merge(ResponseContext.Key.UNCOVERED_INTERVALS_OVERFLOWED, uncoveredIntervalsOverflowed);
```
it is so right now. the merge is applicable only for resulting values

esevastyanov commented on processing/src/main/java/org/apache/druid/query/context/ResponseContext.java:
```diff
@@ -76,56 +138,128 @@ public static ResponseContext createEmpty()
+        responseContext,
+        JacksonUtils.TYPE_REFERENCE_MAP_STRING_OBJECT
+    );
+    return new ResponseContext()
```
Resulting ResponseContext depends on a TypeReference so in general in case of changing the TypeReference the resulting context should be also updated. To eliminate the possible resulting context update I used the inner class. If that fits I can add this description as a comment, if not I may remove inner class usage and use DefaultReponseContext as a resulting map.

esevastyanov commented on processing/src/main/java/org/apache/druid/query/context/ResponseContext.java:
```diff
@@ -76,56 +138,128 @@ public static ResponseContext createEmpty()
+  /**
+   * Serializes the context given that the resulting string length is less than the provided limit.
+   * The method removes max-length fields one by one if the resulting string length is greater than the limit.
+   * The resulting string might be correctly deserialized as a {@link ResponseContext}.
```
Previously the serialized value was literally truncated (hence non-deserializable) thus, in my opinion, there is no correct solution on how to reduce the size of the serialized value.
+ Manually specified priorities of keys. Ideally, this method requires some research on how context values are used by users (because it is a part of the response). E.g. from codebase perspective, it's known that `Etag` key is for caching results as a part HTTP protocol. However, it's not obvious how other keys (values) are used. 
+ Reducing the value of specific keys. Thought about this approach and it seems close to the ideal solution. The only issue is it requires additional information about values (similar to merge function) and much more iterations on reduce phase (in case of using algorithm analogous to the current).

In my view, both of these approaches look as premature optimization without understanding the scope of the issue.

leventov replied:
1) Please put this discussion in code comments.

2) I see a regression scenario: before this PR, `UNCOVERED_INTERVALS` and `MISSING_SEGMENTS` keys were always reasonably short. Now, they may grow very large at Broker, and Broker will prune them altogether. I suggest to hard-code reduction logic specifically for `UNCOVERED_INTERVALS` and `MISSING_SEGMENTS`.

pjain1 replied:
I don't remember exactly but some of the systems relied on having the `MISSING_SEGMENTS` key in the header to do something, so removing the entire entry would break the logic for them. @will-lauer can confirm ? 

I agree a good solution would be to have a `reduce_length` function in the enum itself which would reduce the length step by step (for example removing segment information one by one for missing segments key) until the header length is in bounds. Probably it can be skipped because as per my understanding the truncation previously was random without any guarantees on what keys will be present or truncated. @gianm @himanshug any thoughts on this ?

esevastyanov replied:
[Update the logic of truncation]( https://github.com/apache/incubator-druid/pull/8157#issuecomment-517034018)

esevastyanov commented on processing/src/main/java/org/apache/druid/query/context/ResponseContext.java:
```diff
@@ -21,51 +21,113 @@
+     * Entity tag. A part of HTTP cache validation mechanism.
+     * Is being removed from the context before sending and used as a separate HTTP header.
+     */
+    ETAG("ETag"),
```
`ETag` is a widely used [term](https://en.wikipedia.org/wiki/HTTP_ETag) in the context of HTTP protocol.

esevastyanov commented on processing/src/main/java/org/apache/druid/query/context/ResponseContext.java:
```diff
@@ -76,56 +138,128 @@ public static ResponseContext createEmpty()
+   * The method removes max-length fields one by one if the resulting string length is greater than the limit.
+   * The resulting string might be correctly deserialized as a {@link ResponseContext}.
+   */
+  public SerializationResult serializeWith(ObjectMapper objectMapper, int maxLength) throws JsonProcessingException
```
renamed the argument (also units are mentioned in a method description)

esevastyanov commented on processing/src/main/java/org/apache/druid/query/context/ResponseContext.java:
```diff
@@ -76,56 +138,128 @@ public static ResponseContext createEmpty()
+    } else {
+      final HashMap<String, Object> copiedMap = new HashMap<>(getDelegate());
+      final PriorityQueue<Pair<String, String>> serializedPairs = new PriorityQueue<>((o1, o2) -> {
+        Preconditions.checkNotNull(o1.rhs);
```
removed the checks

esevastyanov commented on processing/src/main/java/org/apache/druid/query/context/ResponseContext.java:
```diff
@@ -76,56 +138,128 @@ public static ResponseContext createEmpty()
+      return new SerializationResult(fullSerializedString, fullSerializedString, false);
+    } else {
+      final HashMap<String, Object> copiedMap = new HashMap<>(getDelegate());
+      final PriorityQueue<Pair<String, String>> serializedPairs = new PriorityQueue<>((o1, o2) -> {
```
great, it's much shorter. replaced

esevastyanov commented on processing/src/main/java/org/apache/druid/query/context/ResponseContext.java:
```diff
@@ -76,56 +138,128 @@ public static ResponseContext createEmpty()
+        return o2.rhs.length() - o1.rhs.length();
+      });
+      for (Map.Entry<String, Object> e : copiedMap.entrySet()) {
+        serializedPairs.add(new Pair<>(e.getKey(), objectMapper.writeValueAsString(e.getValue())));
```
Replaced with the combination of types you listed

esevastyanov commented on processing/src/main/java/org/apache/druid/query/context/ResponseContext.java:
```diff
@@ -76,56 +138,128 @@ public static ResponseContext createEmpty()
+        final Pair<String, String> maxLengthPair = serializedPairs.poll();
+        Preconditions.checkNotNull(maxLengthPair);
+        copiedMap.remove(maxLengthPair.lhs);
+        final String reducedSerializedString = objectMapper.writeValueAsString(copiedMap);
```
Commented

esevastyanov commented on processing/src/main/java/org/apache/druid/query/context/ResponseContext.java:
```diff
@@ -76,56 +138,128 @@ public static ResponseContext createEmpty()
+      }
+      while (!copiedMap.isEmpty()) {
+        final Pair<String, String> maxLengthPair = serializedPairs.poll();
+        Preconditions.checkNotNull(maxLengthPair);
```
Did and removed the check

esevastyanov commented on processing/src/main/java/org/apache/druid/query/context/ResponseContext.java:
```diff
@@ -76,56 +138,128 @@ public static ResponseContext createEmpty()
+
+    public Boolean isReduced()
+    {
+      return isReduced;
```
Replaced with the expression

esevastyanov commented on processing/src/main/java/org/apache/druid/query/context/ResponseContext.java:
```diff
@@ -76,56 +138,128 @@ public static ResponseContext createEmpty()
-        JacksonUtils.TYPE_REFERENCE_MAP_STRING_OBJECT
-    );
-    return new ResponseContext()
+    private final String result;
```
Renamed

esevastyanov commented on processing/src/main/java/org/apache/druid/query/scan/ScanQueryEngine.java:
```diff
@@ -67,14 +67,14 @@
 
-    if (responseContext.get(ResponseContext.CTX_COUNT) != null) {
-      long count = (long) responseContext.get(ResponseContext.CTX_COUNT);
+    if (responseContext.get(ResponseContext.Key.COUNT) != null) {
```
Updated

esevastyanov commented on processing/src/main/java/org/apache/druid/query/context/ResponseContext.java:
```diff
@@ -21,51 +21,113 @@
+    /**
+     * The number of scanned rows.
+     */
+    COUNT(
```
Renamed and left a comment

esevastyanov commented on processing/src/main/java/org/apache/druid/query/context/ResponseContext.java:
```diff
@@ -21,51 +21,113 @@
+    /**
+     * Query fail time (current time + timeout).
+     */
+    QUERY_FAIL_TIME("queryFailTime"),
```
Renamed and updated the descriptions

esevastyanov commented on processing/src/main/java/org/apache/druid/query/context/ResponseContext.java:
```diff
@@ -21,51 +21,113 @@
+     * This variable indicates when a running query should be expired,
+     * and is effective only when 'timeout' of queryContext has a positive value.
+     */
+    TIMEOUT_AT("timeoutAt"),
```
Commented

esevastyanov commented on processing/src/main/java/org/apache/druid/query/context/ResponseContext.java:
```diff
@@ -21,51 +21,113 @@
+    /**
+     * CPU consumed while processing a request.
+     */
+    CPU_CONSUMED(
```
Renamed and commented

esevastyanov commented on processing/src/main/java/org/apache/druid/query/scan/ScanQueryEngine.java:
```diff
@@ -121,9 +121,7 @@
-    if (responseContext.get(ResponseContext.CTX_COUNT) == null) {
-      responseContext.put(ResponseContext.CTX_COUNT, 0L);
-    }
+    responseContext.merge(ResponseContext.Key.COUNT, 0L);
```
Renamed `merge` to `add` (works for both numbers and lists)

esevastyanov commented on processing/src/main/java/org/apache/druid/query/context/ResponseContext.java:
```diff
@@ -21,51 +21,113 @@
-   * The number of scanned rows.
-   */
-  public static final String CTX_COUNT = "count";
+  public enum Key
```
Developed extension-friendly enum with an example

leventov commented on processing/src/main/java/org/apache/druid/query/context/ResponseContext.java:
```diff
@@ -84,36 +125,81 @@
-     * Merge function associated with a key: Object (Object oldValue, Object newValue)
+     * TreeMap is used to have the natural ordering of its keys
      */
+    private static Map<String, BaseKey> map = new TreeMap<>();
```
Suggested to call it "registeredKeys".

leventov replied:
I think this static variable and the associated methods don't need to be nested in `Key`. They might as well be in the higher-level `ResponseContext`.

esevastyanov replied:
Yes, they might be there. But I think we may leave them inside enum as this static variable and methods are part of enum "extension" and might be helpful in a way of understanding how this "extension" is implemented. Since there is no built-in enum extension support this implementation may be used as an example in some cases so I would prefer to keep enum and this static field and methods together.

leventov commented on processing/src/main/java/org/apache/druid/query/context/ResponseContext.java:
```diff
@@ -84,36 +125,81 @@
+     * The primary way of registering context keys.
+     * Only the keys registered this way are considered during the context merge.
+     */
+    public static void addKey(BaseKey key)
```
What about "registerKey"?

esevastyanov replied:
renamed

leventov commented on processing/src/main/java/org/apache/druid/query/context/ResponseContext.java:
```diff
@@ -84,36 +125,81 @@
+
+    /**
+     * The primary way of registering context keys.
+     * Only the keys registered this way are considered during the context merge.
```
Please note what happens if a context has an unregistered key. (I think, ideally, it should throw ISE.)

esevastyanov replied:
Updated exceptions and related comments

leventov commented on processing/src/main/java/org/apache/druid/query/context/ResponseContext.java:
```diff
@@ -84,36 +125,81 @@
     ETAG("ETag"),
     /**
      * Query fail time (current time + timeout).
+     * The final value in comparison to continuously updated TIMEOUT_AT.
```
I failed to understand this sentence after several readings. Please reword.

esevastyanov replied:
reworded it

leventov commented on processing/src/main/java/org/apache/druid/query/context/ResponseContext.java:
```diff
@@ -44,9 +47,47 @@
+   *   @Override public BiFunction<Object, Object, Object> getMergeFunction() { return mergeFunction; }
+   * }
+   * }</pre>
+   * Make sure all extension enum values added with Key.addKey method.
```
Please make `Key.addKey` a `{@link }`

esevastyanov replied:
Updated

leventov commented on processing/src/main/java/org/apache/druid/query/context/ResponseContext.java:
```diff
@@ -44,9 +47,47 @@
+  }
+
+  /**
+   * Keys associated with objects in the context. The enum is extension-friendly.
```
I think it doesn't make a lot of sense to say that "The enum is extension-friendly." Enum is not extension-friendly, but the key system (based from `BaseKey`) is. So I would just remove this sentence.

esevastyanov replied:
Removed

leventov commented on processing/src/main/java/org/apache/druid/query/context/ResponseContext.java:
```diff
@@ -84,36 +125,81 @@
+    /**
+     * Returns all keys the enum contains and the added via addKey method
+     */
+    public static Collection<BaseKey> getKeys()
```
Suggested "getAllRegisteredKeys"

esevastyanov replied:
Renamed

leventov commented on processing/src/main/java/org/apache/druid/query/scan/ScanQuery.java:
```diff
@@ -334,7 +336,7 @@ public String toString()
            ", resultFormat='" + resultFormat + '\'' +
            ", batchSize=" + batchSize +
-           ", limit=" + limit +
+           ", limit=" + scanRowsLimit +
```
There are no backward compatibility concerns in `toString()`, please change the key.

esevastyanov replied:
Changed

will-lauer commented:
@pjain1 While we certainly talked about using MISSING_SEGMENTS, I don't believe we ever actually implemented it in production, so while removing it completely probably won't break anything we have, it is less than desirable. I'd prefer to have a partial list than no list at all. Or at least some other indication that the list was non-empty.

himanshug commented:
> Also improved ResponseContext serialization. Previously the result of serialization was truncated if its length was greater than the limit. I believe it would be better to keep the context structure and make it deserializable thus the process of serialization removes max-length fields completely from the context until the final result length doesn't exceed the limit.

I haven't looked at the code , but this is definitely an incompatibility with previous behavior, so tagged the PR as such . It should also be mentioned in release notes. I don't think most customers would care about it.
Now, if possible, a better strategy might be to retain all keys but trim the bigger ones so to keep serialized response limited.
Or, don't change the behavior since we haven't had use cases where this has been a problem so far.

esevastyanov commented on processing/src/main/java/org/apache/druid/query/context/ResponseContext.java:
```diff
@@ -76,56 +236,129 @@ public static ResponseContext createEmpty()
+        ));
+      }
+      // quadratic complexity: while loop with map serialization on each iteration
+      while (!copiedMap.isEmpty() && !serializedValueEntries.isEmpty()) {
```
Updated the whole method

esevastyanov commented on processing/src/main/java/org/apache/druid/query/scan/ScanQuery.java:
```diff
@@ -125,7 +126,7 @@ public ScanQuery(
       @JsonProperty("resultFormat") ResultFormat resultFormat,
       @JsonProperty("batchSize") int batchSize,
-      @JsonProperty("limit") long limit,
+      @JsonProperty("limit") long scanRowsLimit,
```
Almost unrelated to the PR, just to make [things clearer](https://github.com/apache/incubator-druid/pull/8157#discussion_r307378885)

esevastyanov commented on processing/src/test/java/org/apache/druid/query/topn/TopNQueryRunnerTest.java:
```diff
@@ -1301,9 +1301,7 @@ public void testTopNBySegment()
-    final ResponseContext responseContext = ResponseContext.createEmpty();
-    responseContext.putAll(specialContext);
-    Sequence<Result<TopNResultValue>> results = runWithMerge(query, responseContext);
+    Sequence<Result<TopNResultValue>> results = runWithMerge(query);
```
Yes, according to the test result. In my view, it was a legacy or an incorrect test.

esevastyanov commented:
Thanks, everyone. I updated the logic of truncation and kept it as general as possible without mentioning custom context keys. New algorithm removes some values from resulting (serialized) JSON arrays (no matter if it's `MISSING_SEGMENTS` or `UNCOVERED_INTERVALS`) to satisfy the length limit. So as previously, the serialized context has some but not all array's values if the limit exceeded. I also added a boolean indicator if a context was truncated.

pjain1 commented on processing/src/main/java/org/apache/druid/query/context/ResponseContext.java:
```diff
@@ -76,56 +253,158 @@ public static ResponseContext createEmpty()
+      sortedNodesByLength.sort(valueLengthReversedComparator);
+      int needToRemoveCharsNumber = fullSerializedString.length() - maxCharsNumber;
+      // In general, the complexity of this block is O(n*m*log(m)) where n - context size, m - context's array size
+      for (Map.Entry<String, JsonNode> e : sortedNodesByLength) {
```
nit: can it be simplified to handle array and non-array cases separately like -
```
    for (Map.Entry<String, JsonNode> e : sortedNodesByLength) {
      final String fieldName = e.getKey();
      final JsonNode node = e.getValue();
      if (node.isArray()) {
        if(needToRemoveCharsNumber >= node.toString().length()) {
          final int lengthBeforeRemove = node.toString().length();
          // Empty array could be correctly deserialized so we remove only its elements.
          ((ArrayNode) node).removeAll();
          final int lengthAfterRemove = node.toString().length();
          needToRemoveCharsNumber -= lengthBeforeRemove - lengthAfterRemove;
        } else {
          final ArrayNode arrNode = (ArrayNode) node;
          while (node.size() > 0 && needToRemoveCharsNumber > 0) {
            final int lengthBeforeRemove = arrNode.toString().length();
            // Reducing complexity by removing half of array's elements
            final int removeUntil = node.size() / 2;
            for (int removeAt = node.size() - 1; removeAt >= removeUntil; removeAt--) {
              arrNode.remove(removeAt);
            }
            final int lengthAfterRemove = arrNode.toString().length();
            needToRemoveCharsNumber -= lengthBeforeRemove - lengthAfterRemove;
          }
        } // node is not an array
      } else {
        // In general, a context should not contain nulls so we completely remove the field.
        contextJsonNode.remove(fieldName);
        // Since the field is completely removed (name + value) we need to do a recalculation
        needToRemoveCharsNumber = contextJsonNode.toString().length() - maxCharsNumber;
      }
      if (needToRemoveCharsNumber <= 0) {
        break;
      }
    }
```
I think the logic is still the same but simpler to understand.

esevastyanov replied:
looks nice. updated

leventov commented on processing/src/main/java/org/apache/druid/query/context/ResponseContext.java:
```diff
@@ -155,47 +159,60 @@
+     * @throws IllegalArgumentException if the key has already been registered.
      */
-    public static void addKey(BaseKey key)
+    public static void registerKey(BaseKey key)
```
Just in case, please make this method `synchronized`

esevastyanov replied:
added `synchronized`

leventov replied:
Turns out it was not quite enough, #9106

leventov commented on processing/src/main/java/org/apache/druid/query/context/ResponseContext.java:
```diff
@@ -155,47 +159,60 @@
+    public static Collection<BaseKey> getAllRegisteredKeys()
     {
-      return map.values();
+      return registeredKeys.values();
```
Just in case, please wrap with `Collections.unmodifiableCollection()`

esevastyanov replied:
wrapped

leventov commented on processing/src/main/java/org/apache/druid/query/context/ResponseContext.java:
```diff
@@ -236,95 +253,124 @@ public static ResponseContext createEmpty()
-   * The method removes max-length fields one by one if the resulting string length is greater than the limit.
-   * The resulting string might be correctly deserialized as a {@link ResponseContext}.
+   * This method tries to remove some elements from context collections if it's needed to satisfy the limit.
+   * The resulting string might be correctly deserialized to {@link ResponseContext}.
```
Please comment on why explicit priorities of keys are not implemented.

esevastyanov replied:
commented

leventov commented on processing/src/main/java/org/apache/druid/query/context/ResponseContext.java:
```diff
@@ -236,95 +253,124 @@ public static ResponseContext createEmpty()
+      for (Map.Entry<String, JsonNode> e : sortedNodesByLength) {
+        final String fieldName = e.getKey();
+        final JsonNode node = e.getValue();
+        if (node.isArray()) {
```
If this block aims for `MISSING_SEGMENTS` and `UNCOVERED_INTERVALS`, please comment on that with an example.

esevastyanov replied:
commented

leventov commented on processing/src/main/java/org/apache/druid/query/context/ResponseContext.java:
```diff
@@ -236,95 +253,124 @@ public static ResponseContext createEmpty()
+        if (node.isArray()) {
+          if (needToRemoveCharsNumber >= node.toString().length()) {
+            final int lengthBeforeRemove = node.toString().length();
+            // Empty array could be correctly deserialized so we remove only its elements.
```
I think the logic of this block should avoid producing empty array because it may be misleading.

esevastyanov replied:
Removed empty arrays

leventov commented on processing/src/main/java/org/apache/druid/query/context/ResponseContext.java:
```diff
@@ -236,95 +253,124 @@ public static ResponseContext createEmpty()
+      add(Key.TRUNCATED, true);
+      final ObjectNode contextJsonNode = objectMapper.valueToTree(getDelegate());
+      final ArrayList<Map.Entry<String, JsonNode>> sortedNodesByLength = Lists.newArrayList(contextJsonNode.fields());
+      final Comparator<Map.Entry<String, JsonNode>> valueLengthReversedComparator =
```
Please extract this comparator as a constant.

esevastyanov replied:
Extracted

leventov commented on processing/src/main/java/org/apache/druid/query/context/ResponseContext.java:
```diff
@@ -76,56 +253,158 @@ public static ResponseContext createEmpty()
+            final int lengthAfterRemove = node.toString().length();
+            needToRemoveCharsNumber -= lengthBeforeRemove - lengthAfterRemove;
+          } else {
+            final ArrayNode arrNode = (ArrayNode) node;
```
This block needs a comment. It's not obvious what and why is going on here. Please extract as a method (or the upper block) if possible.

esevastyanov replied:
added a comment and extracted

leventov commented on processing/src/main/java/org/apache/druid/query/context/ResponseContext.java:
```diff
@@ -244,6 +245,11 @@ public String getName()
 
+  protected abstract Map<BaseKey, Object> getDelegate();
+
+  private final Comparator<Map.Entry<String, JsonNode>> valueLengthReversedComparator =
```
It can be static final constant.

esevastyanov replied:
updated

leventov commented on processing/src/main/java/org/apache/druid/query/context/ResponseContext.java:
```diff
@@ -319,7 +323,11 @@ public void merge(ResponseContext responseContext)
-   * This method tries to remove some elements from context collections if it's needed to satisfy the limit.
+   * This method removes some elements from context collections if it's needed to satisfy the limit.
+   * There is no explicit priorities of keys which values are being truncated because for now there are only
+   * two potential limit breaking keys (UNCOVERED_INTERVALS and MISSING_SEGMENTS) and their values are arrays.
```
Please wrap `UNCOVERED_INTERVALS` and `MISSING_SEGMENTS` with `{@link }`

esevastyanov replied:
wrapped

leventov commented on processing/src/main/java/org/apache/druid/query/context/ResponseContext.java:
```diff
@@ -342,26 +348,22 @@ public SerializationResult serializeWith(ObjectMapper objectMapper, int maxChars
-            ((ArrayNode) node).removeAll();
-            final int lengthAfterRemove = node.toString().length();
-            needToRemoveCharsNumber -= lengthBeforeRemove - lengthAfterRemove;
+            // We need to remove more chars than the field's lenght so removing it completely
```
Typo, "length". There is one other instance of this typo in the repository, in `StringDimensionHandler` - please fix it too.

esevastyanov replied:
fixed both typos

leventov commented on processing/src/main/java/org/apache/druid/query/context/ResponseContext.java:
```diff
@@ -374,6 +376,30 @@ public SerializationResult serializeWith(ObjectMapper objectMapper, int maxChars
   }
 
+  /**
+   * Removes {@code node}'s elements which total lenght of serialized values is greater or equal to the passed limit.
```
Same

esevastyanov replied:
fixed

leventov commented on processing/src/main/java/org/apache/druid/query/context/ResponseContext.java:
```diff
@@ -374,6 +376,30 @@ public SerializationResult serializeWith(ObjectMapper objectMapper, int maxChars
+   * @param needToRemoveCharsNumber the number of chars need to be removed.
+   * @return the number of removed chars.
+   */
+  private int removeNodeElementsToSatisfyCharsLimit(ArrayNode node, int needToRemoveCharsNumber)
```
Looks like this method can be static.

esevastyanov replied:
updated

leventov commented on processing/src/main/java/org/apache/druid/query/context/ResponseContext.java:
```diff
@@ -342,26 +348,22 @@ public SerializationResult serializeWith(ObjectMapper objectMapper, int maxChars
+            final ArrayNode arrayNode = (ArrayNode) node;
+            needToRemoveCharsNumber -= removeNodeElementsToSatisfyCharsLimit(arrayNode, needToRemoveCharsNumber);
+            if (arrayNode.size() == 0) {
+              // The field is empty, removing it.
```
Please extend the comment like `The field is empty, removing it because an empty array field may be misleading for the recipients of the truncated response context.`

esevastyanov replied:
Extended

leventov commented on processing/src/main/java/org/apache/druid/query/context/ResponseContext.java:
```diff
@@ -19,53 +19,236 @@
+    ETAG("ETag"),
+    /**
+     * Query fail time (current time + timeout).
+     * It is not updated continuously as TIMEOUT_AT.
```
Please wrap `TIMEOUT_AT` with `{@link }`.

esevastyanov replied:
wrapped
